ReactiveCocoa/DelegateProxy.swift

```diff
@@ -1,106 +1,269 @@
 import ReactiveSwift
 import enum Result.NoError
 
-internal class DelegateProxy<Delegate: NSObjectProtocol>: NSObject {
-	internal weak var forwardee: Delegate? {
-		didSet {
-			originalSetter(self)
-		}
+private let delegateProxySetupKey = AssociationKey<Bool>(default: false)
+private let hasSwizzledKey = AssociationKey<Bool>(default: false)
+
+public protocol DelegateProxyProtocol: class {}
+
+// This is supposedly private, and is made `internal` to circumvent a serializer bug.
+internal protocol _DelegateProxyProtocol: class {
+	var _forwardee: AnyObject? { get }
+	func proxyWillIntercept(_ selector: Selector)
+	var lifetime: Lifetime { get }
+}
+
+public struct DelegateProxyConfiguration {
+	fileprivate let objcProtocol: Protocol
+	fileprivate let lifetime: Lifetime
+	fileprivate let originalSetter: (AnyObject) -> Void
+}
+
+open class DelegateProxy<Delegate: NSObjectProtocol>: NSObject, DelegateProxyProtocol, _DelegateProxyProtocol {
```

Comment:
Why make this `open`? Why do you envision subclassing it?

Reply:
> Subclassing is no longer necessary, but can still be used to supply non-void-returning implementations.

Method interception only supports `<U...>(_ args: U...) -> Void`, but not `<U..., T>(_ args: U...) -> T`.

Reply:
This seems like it gives only very small gains for quite a bit of effort.

Since it's not automatic, it doesn't seem any better than setting the `delegate` yourself and using a custom delegate alongside `DelegateProxy`. The type inference on `Reactive.proxy(forKey:)` is also a bit problem since the `DelegateProxy` subclass is only specified in the return type.

I think we should leave this `final` and remove the generics.